README.md
# grunt-jira-actions

> Grunt tasks to perform Jira actions.

[![npm version](https://badge.fury.io/js/grunt-jira-actions.svg)](http://badge.fury.io/js/grunt-jira-actions)
![Dependencies](https://david-dm.org/jwtd/grunt-jira-actions.png)
[![Build Status](https://travis-ci.org/jwtd/grunt-jira-actions.png)](https://travis-ci.org/jwtd/grunt-jira-actions)
[![Inline docs](http://inch-ci.org/github/jwtd/grunt-jira-actions.svg?branch=master)](http://inch-ci.org/github/jwtd/grunt-jira-actions)
[![NPM](https://nodei.co/npm/grunt-jira-actions.png)](https://nodei.co/npm/grunt-jira-actions)

## What You Can Do
* [Specify Jira connection properties](#setJiraConfig)
* [Create any type of Jira issue (Epic, Story, Bug, etc)](#createJiraIssue)
* [Transition issues to any state (Open, In Development, Closed, etc)](#transitionJiraIssue)
* [Link two issues together (Blocks, Cloners, Duplicate, Relates)](#linkJiraIssues)
* [Add a comment to an existing issue](#addJiraComment)
* [Create a new Version in a Jira project](#createJiraVersion)
* [Fetch a hash of Jira issues using a JQL search](#searchJira)


### Requirements

* Grunt `>=0.4.0`
* API access to a Jira instance

### Installation

I bet you were expecting me to to tell you that if you haven't used [Grunt](http://gruntjs.com/) before, to be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, because it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins.

Well, I'm not going to do that. Instead, I'm going to assume you are familiar with that process, and will skip ahead to explaining how you can install this plugin. You do so with this command:

```shell
npm install grunt-jira-actions --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-jira-actions');
```

# Performing Actions in Jira

## <a name="setJiraConfig"></a>Specify Jira Connection Properties

The following options are used by all Jira Action tasks:

- `env_var_for_jira_username` - Environment variable that holds Jira username. Default is 'JIRA_UN'.
- `env_var_for_jira_password` - Environment variable that holds Jira password. Default is 'JIRA_PW'.
- `jira_host` - Base domain of your Jira instance's api root (i.e. 'foo.atlassian.net').
- `jira_protocol` - The protocol which Jira's api uses for connections. Default is 'https'.
- `jira_port` - The port on which Jira's api allows connections on. Default is 443.
- `jira_api_version` - The version of Jira's api to target. Default is '2'.

These values can be set:
* Globaly using the `setJiraConfig` task
* In each task's top level options collection (will override global values)
* In each target's option collection (will override task and global values)

### Example

```js

module.exports = function(grunt) {

  grunt.initConfig({

    // These values will be the defaults used in all Jira action tasks and targets
    setJiraConfig: {
      options: {
        env_var_for_jira_username: 'JIRA_UN',
        env_var_for_jira_password: 'JIRA_PW',
        jira_host: 'foo.atlassian.net',
        jira_protocol: 'https',
        jira_port: 443,
        jira_api_version: '2'
      }
    },

    // Multi-task for creating Jira stories
    createJiraIssue: {

      // Override global Jira config options with task specific options
      options: {
        env_var_for_jira_username: 'BAR_JIRA_UN',
        env_var_for_jira_password: 'BAR_JIRA_PW',
        jira_host: 'bar.atlassian.net',
        jira_protocol: 'http',
        jira_port: 80,
        jira_api_version: '2'
      },

      // Override global and task specific Jira config options with target specific options
      createFooStory: {
        options: {
          env_var_for_jira_username: 'BAZ_JIRA_UN',
          env_var_for_jira_password: 'BAZ_JIRA_PW',
          jira_host: 'baz.atlassian.net',
          jira_protocol: 'http',
          jira_port: 80,
          jira_api_version: '2'
        }
      }

    }

    // Load the task
    grunt.loadNpmTasks('grunt-jira-actions');

    // Call setJiraConfig, before any other Jira action tasks
    grunt.registerTask('default', ['setJiraConfig', 'createJiraIssue:createFooStory']);

  });
}

```

## <a name="createJiraIssue"></a>Creating Jira Issues with 'createJiraIssue'

In your project's Gruntfile, add a section named `createJiraIssue` to the data object passed into `grunt.initConfig()`. Within that section, you can create any number of targets that add Jira issues of various types. Place common values in the top level options collection. Place target specific values in their respective target's option's collections.

### Parameters specific to `createJiraIssue` targets

- `project_id` - Jira id of the project the story will be created in.
- `issue_type` - Jira name of the type of issue to be created. Default is 'Story'. Valid values are:
 - Bug
 - New Feature
 - Task
 - Improvement
 - Sub-task
 - Epic
 - Story
 - Technical Task
- `issue_state` - The transition id that the issue should end up in. Default is 1 which is Open. 2 is Closed.
- `summary` - Default is the project name specified in package.json (displayed in the story's subject)
- `description` - The description of the issue being created. If value is a valid file path, the contents of the file will be used (plain txt and JSON are supported).
- `optional_fields` - JSON to be added to the create issue call's fields JSON. For more details check [developer.atlassian.com](https://developer.atlassian.com/display/JIRADEV/JIRA+REST+API+Example+-+Create+Issue)

### Example

```js

grunt.initConfig({


    createJiraIssue: {

      // Declare options that are common to all Jira actions (or call setJiraConfig task before this one)
      options: {
        jira_host: 'bar.atlassian.net',
        project_id: 123456              // This will be common to all targets
      },

      // Create specific targets to perform different Jira tasks
      createAndCloseFooStory: {
        options: {
          issue_type: 'Story', // Story, Epic, Task, Technical Task, Sub-Task, Bug, Improvement, New Feature
          issue_state: 1,      // 1 = Open, 2 = Done
          summary: 'This is the foo story summary',
          description: 'This is the foo story description.'
        }
      },

      // Create specific targets to perform different Jira tasks
      createOpenBarTask: {
        options: {
          issue_type: 'Task',
          issue_state: 2,      // 1 = Open, 2 = Done
          summary: 'This is the bar task summary',
          description: 'test/data/issue_body.txt',  // Import contents of file as description
          optional_fields: {
            'priority': {
              'name': 'Major'  // Specify priority - Critical, Major, Medium (default), Minor
            },
            'components': [{
              'id': '56789'    // Specify a component
            }]
          }
        }
      }

    }

});

```

## <a name="transitionJiraIssue"></a>Transition an Existing Jira Issue 'transitionJiraIssue'

The `transitionJiraIssue` task is called from other tasks, but can also be called directly using the format `grunt transitionJiraIssue:<issue_id>:<issue_state>`.

### Parameters specific to `createJiraIssue` targets

- `issue_id` - Jira id of the issue that will be transitioned.
- `issue_state` - The transition id that the issue should end up in. Default is 1 which is Open. 2 is Closed.

### Example

```shell

grunt transitionJiraIssue:19416:2

```


## <a name="linkJiraIssues"></a>Add Comments to Existing Issues with 'linkJiraIssues'

The `linkJiraIssues` task can be called directly using the format `grunt linkJiraIssues:<from_issue_key>:<to_issue_key>:<link_type>:<comment>`.

### Parameters specific to `linkJiraIssues` target

- `from_issue_key` - Jira issue KEY that will be linked.
- `to_issue_key` - Jira issue KEY that will be linked.
- `link_type` - The type of link that should be setup. Valid values are 'Blocks', 'Cloners', 'Duplicate', 'Relates'. Default is 'Relates'.
- `comment` - The body of the comment that should accompany the link. If value is a valid file path, the contents of the file will be used (plain txt and JSON are supported).


### Example

```shell

grunt linkJiraIssues:GEN-200:GEN-201:Relates:tmp/link/description.txt

```


## <a name="addJiraComment"></a>Add Comments to Existing Issues with 'addJiraComment'

In your project's Gruntfile, add a section named `addJiraComment` to the data object passed into `grunt.initConfig()`. Within that section, you can create any number of targets that will add Jira comments to existing issues. Place common values in the top level options collection. Place target specific values in their respective target's option's collections.

### Parameters specific to `addJiraComment` target

- `issue_id` - Jira id of the project the story will be created in.
- `comment` - The body of the comment being added. If value is a valid file path, the contents of the file will be used (plain txt and JSON are supported).

### Example

```js

grunt.initConfig({

    // Add comments to existing Jira issues
    addJiraComment: {

      // Declare options that are common to all Jira actions (or call setJiraConfig task before this one)
      options: {
        jira_host: 'foo.atlassian.net'
      },

      // Create specific targets to perform different Jira tasks
      onIssue: {
        options: {
          // issue_id: This value will be passed in via target call such as addJiraComment:onIssue:19400
          comment: 'This is a comment on the story.'
        }
      },

      // Create specific targets to perform different Jira tasks
      fromFileToIssue: {
        options: {
          // issue_id: This value will be passed in via target call such as addJiraComment:fromFileToIssue:19400
          comment: 'test/data/comment_body.txt'
        }
      }

    }

});

```


## <a name="createJiraVersion"></a>Create a Project Version with 'createJiraVersion'

In your project's Gruntfile, add a section named `createJiraVersion` to the data object passed into `grunt.initConfig()`. Within that section, you can create any number of targets that can be used to create project versions. Place common values in the top level options collection. Place target specific values in their respective target's option's collections.

### Parameters specific to `createJiraVersion` target

- `project` - Jira's three letter project key (ex. 'FOO')
- `name` - String that will be used as the name of the version.
- `description` - The description of the Version being created. If value is a valid file path, the contents of the file will be used (plain txt and JSON are supported).
- `archived` - Boolean value indicating whether the version has been archived. Default is false.
- `released` - Boolean value indicating whether the version has been released. Default is false.
- `release_date` - Date on which the version was released. Default is the current date.

### Example

```js

grunt.initConfig({

    // Create a Version of a Jira project
    createJiraVersion: {

      // Declare options that are common to all Jira actions (or call setJiraConfig task before this one)
      options: {
        jira_host: 'foo.atlassian.net'
      },

      // Create specific targets for different Jira projects
      fooProject: {
        options: {
          project_key: 'FOO',
          name: 'New Version 1',
          description: 'This is the next version.',
          archived: false,
          released: true,
          release_date: '2015-02-21'
        }
      },

      // Create specific targets for different Jira projects
      barProject: {
        options: {
          project_key: 'BAR',
          name: 'New Version 1',
          description: 'test/data/version_description.txt',
          archived: false,
          released: true,
          release_date: '2015-02-21'
        }
      },
    }

});

```


## <a name="searchJira"></a>Use JQL to fetch a hash of Jira issues with 'searchJira'

In your project's Gruntfile, add a section named `searchJira` to the data object passed into `grunt.initConfig()`. Within that section, you can create any number of targets that can be used to submit different JQL queries. Place common values in the top level options collection. Place target specific values in their respective target's option's collections.

Each target within this task will save its results `grunt.config` with the variable `this.target + 'search_results` as a hash of issues with the issue key as the hash key.

### Parameters specific to `searchJira` target

- `search_string` - A valid JQL search string.
- `start_at` - Optional page offset for the search. Default is 0.
- `max_results` - Optional max limit of issues. Default is 9999.

### Example

```js

grunt.initConfig({

    // Search Jira for specific issues
    searchJira: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'foo.atlassian.net'
      },

      // Create specific targets for different Jira searches
      // This target's search results can be accessed via grunt.config('forGenStoriesAndBugs.search_results');
      forGenStoriesAndBugs: {
        options: {
          search_string: 'project="GEN" AND status="OPEN" AND issuetype in ("Bug","Story")',
          //start_at: 0,
          //max_results: 3,
        }
      }

      // Add another JQL search
      // This target's search results can be accessed via grunt.config('forGenEpics.search_results');
      forGenEpics: {
        options: {
          search_string: 'project="GEN" issuetype in ("Epic")',
        }
      }

    }

});

```


### JQL Search Results

The results hash contains the issue id, key, summary, status id, and status label.

```

{
  'GEN-1':
   { id: '10221',
     key: 'GEN-1',
     summary: 'Issue summary foo',
     status_id: '1',
     status: 'Open' },
  'GEN-2':
   { id: '13897',
     key: 'GEN-2',
     summary: 'Issue summary bar',
     status_id: '1',
     status: 'Open' },
  'GEN-3':
   { id: '17429',
     key: 'GEN-3',
     summary: 'Issue summary baz',
     status_id: '1',
     status: 'Open' }
}

```

## Experimental : jiraProjectRapidView

```js

grunt.initConfig({

    // Get a Jira project's rapid view
    jiraProjectRapidView: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'foo.atlassian.net'
      },

      // Create specific targets for different Jira searches
      ofGeneralProject: {
        options: {
          project_key: 'GEN'
        }
      }

    }

});

```

## Contact, feedback and bugs

This interface was not developed or reviewed by Atlassian. They bare no responsibility for its quality, performance, or results. Use at your own risk.

Please file bugs / issues and feature requests on the [issue tracker](https://github.com/jwtd/grunt-jira-actions/issues)

## Contributing
The code styleguide for this project is captured in the [eslint.json](https://github.com/jwtd/grunt-jira-actions/blob/master/eslint.json) file. Submissions should be accompanied by unit tests for any new or changed functionality. [esLint](http://eslint.org/) and test your code using [Grunt](http://gruntjs.com/).

## Special Thanks
* Ryan Tomlinson for the helpful write up of [OpenTable's release process](http://tech.opentable.co.uk/blog/2014/05/19/continuous-delivery-automating-deployment-visibility/)
* [OpenTable](https://github.com/opentable) - for being a progressive organization and allowing their staff to open source their tools.
* [Chris Riddle](https://github.com/christriddle) and [Bryce Catlin](https://github.com/bcatlin) for putting together [grunt-ccb](https://github.com/opentable/grunt-ccb), which this project is derived from.


CHANGELOG.md
### 1.0.5 - 2016-03-29
### 1.0.4 - 2016-03-29
### 1.0.2 - 2015-04-23
### 1.0.1 - 2015-04-23
## Change Log

### upcoming (2015/04/09 22:03 +00:00)
- [e37694e](https://github.com/jwtd/grunt-jira-actions/commit/e37694edd666d3ebc208a890c630e729f72f26ce) Added grunt-github-manifest (@jwtd)
- [624afb5](https://github.com/jwtd/grunt-jira-actions/commit/624afb5a3d48773349f8354a85655b2f0590d2ce) Updated readme with exp task (@jwtd)
- [e084dd9](https://github.com/jwtd/grunt-jira-actions/commit/e084dd9fa795787b43fdaf7a005a955259d9c34a) Added dependencies check badge (@jwtd)
- [d23e18e](https://github.com/jwtd/grunt-jira-actions/commit/d23e18e9a944363825310a82232bbbdda9f3e571) Updated readme (@jwtd)
- [c4fb3b8](https://github.com/jwtd/grunt-jira-actions/commit/c4fb3b87deebd5b84bff4eaf287429499d4f0338) Added npm meta block to readme (@jwtd)
- [bc3b23c](https://github.com/jwtd/grunt-jira-actions/commit/bc3b23c12938973b1b336a8d580d4baf9d4827f9) Added npm version and travis ci build status badges (@jwtd)
- [740fd08](https://github.com/jwtd/grunt-jira-actions/commit/740fd08d2dc88040a9b6941437d1ee43ed197df5) Removed uneeded dependencies to resolve travis build errors (@jwtd)
- [5a09a45](https://github.com/jwtd/grunt-jira-actions/commit/5a09a459d14e56c50a6d26f8c8d6f898618bf1cb) Added travis ci config file (@jwtd)
- [9d8ff18](https://github.com/jwtd/grunt-jira-actions/commit/9d8ff18b163857746e92a06d58dbf8bf74bd44e6) Formatted titles in readme (@jwtd)

### 1.0.0 (2015/03/31 03:17 +00:00)
- [98e2047](https://github.com/jwtd/grunt-jira-actions/commit/98e2047831c90d9010ecd90ef52104cbb0cc36f0) Release of 1.0.0 (@jwtd)
- [7b0975d](https://github.com/jwtd/grunt-jira-actions/commit/7b0975dbc2eca6c8b7da9baea8885d9ac7f960b8) Resolved eslint warnings (@jwtd)
- [fd5f8da](https://github.com/jwtd/grunt-jira-actions/commit/fd5f8dac0b6fbd8b4f2752d52c9422b45257d5bb) Removed code coverage config (@jwtd)
- [e5f9a75](https://github.com/jwtd/grunt-jira-actions/commit/e5f9a75cf0a06e6a9559326c69dd3816bd093d35) Cleaned up comments (@jwtd)
- [0d54462](https://github.com/jwtd/grunt-jira-actions/commit/0d5446209ffd0b41d3e060fcc9dfa2fee026d168) Added jiraProjectRapidView task and cleaned up verbose output when not testing (@jwtd)
- [f95ea3a](https://github.com/jwtd/grunt-jira-actions/commit/f95ea3aacd82f8aa0b30e7278bcaf89520995d05) Updated gruntfile to load instrumented version for code coverage (@jwtd)
- [7da2a6a](https://github.com/jwtd/grunt-jira-actions/commit/7da2a6a83b5cce32d05b9fc3c4bb98b5f9abe217) Removed ENV step in coverage task and replaced it with task import from coverage directory (@jwtd)
- [45c678a](https://github.com/jwtd/grunt-jira-actions/commit/45c678a9e3c65b53ea5b8c3381d663655cabd430) Added require helper (@jwtd)
- [6236a90](https://github.com/jwtd/grunt-jira-actions/commit/6236a909441fccaa79e76b76727211e7a52409e2) Simplified nock recorder and got all test using nock http calls (@jwtd)
- [5814c7e](https://github.com/jwtd/grunt-jira-actions/commit/5814c7efb2ff6656646d92ef5c5afee0a2e0a171) Added grunt-istanbul and supporting modules (@jwtd)
- [e9bdff8](https://github.com/jwtd/grunt-jira-actions/commit/e9bdff8a5268d16b09f2b2ecc018ba7e15f9daa4) Moved init code for tests into helper.js (@jwtd)
- [831eab9](https://github.com/jwtd/grunt-jira-actions/commit/831eab9988f68c2090fcd14768a0bcc916c316a6) Have all tests for creating issues and adding comments working (@jwtd)
- [0b6821f](https://github.com/jwtd/grunt-jira-actions/commit/0b6821ffb1e9be74176831a1f8398e4ce0984824) Setup addJiraComment_tests (@jwtd)
- [cfe80ec](https://github.com/jwtd/grunt-jira-actions/commit/cfe80ec6970cbc610cd9396e676689d65f3e284a) Added createJiraIssue_withInvalidIssueState_should_FAIL and createJiraIssue_withInvalidPriority_should_FAIL tests. (@jwtd)
- [c46ab96](https://github.com/jwtd/grunt-jira-actions/commit/c46ab96560938b2d1bb9c32a62a30b8e223b90e5) Created function to unpack node-jira errors and added createJiraIssue_withInvalidIssueType_should_FAIL (@jwtd)
- [6dc06da](https://github.com/jwtd/grunt-jira-actions/commit/6dc06da78acedba32c43f4ce027bdf44570e47b7) Added failure test createJiraIssue_withoutRequiredOptions_should_FAIL (@jwtd)
- [ec78e3e](https://github.com/jwtd/grunt-jira-actions/commit/ec78e3e53862375ae5fb162b1f8232defd924c3c) Added createJiraIssue_asValidStoryWithPriority_should_PASS (@jwtd)
- [57dae67](https://github.com/jwtd/grunt-jira-actions/commit/57dae677d9d2e3bc25188a2501611db04b5cad9e) Got createJiraIssue_asValidStoryMarkedDone_should_PASS test working. (@jwtd)
- [919c0a1](https://github.com/jwtd/grunt-jira-actions/commit/919c0a1a87e2b0e479b7e864c2c23d057a629f5c) Got tests in createJiraIssue_asValidTaskDescrFromFile_should_PASS pass (@jwtd)
- [dc7f45b](https://github.com/jwtd/grunt-jira-actions/commit/dc7f45b9914749e7fb3476b4d8a37b0f8b623101) Got tests for createJiraIssue_asValidStoryDescrFromOption_should_PASS.js to pass (@jwtd)
- [7ed3148](https://github.com/jwtd/grunt-jira-actions/commit/7ed3148b85c84a9e42fa4d83751bde82609c0cc2) Got all tests of createJiraIssue:withMinimumOptions_should_PASS to pass. (@jwtd)
- [2f69bb5](https://github.com/jwtd/grunt-jira-actions/commit/2f69bb5a535aefd2fb45c1fe91476d56a4cfb2b9) Formatted output of tasks so that they can be parsed back into json from stdout (@jwtd)
- [90c123d](https://github.com/jwtd/grunt-jira-actions/commit/90c123d85beb0409a03968870395e14774ac0aba) Added more test configs for createJiraIssue task (@jwtd)
- [cf52424](https://github.com/jwtd/grunt-jira-actions/commit/cf52424cd42e6a1b0e466ba15c5360677fd298b8) Cleaned up more eslint errors and warnings (@jwtd)
- [f80c14f](https://github.com/jwtd/grunt-jira-actions/commit/f80c14fd3649d2d68f2f3a5a2dd22b74b3016274) Cleaned up eslint errors (@jwtd)
- [5d49073](https://github.com/jwtd/grunt-jira-actions/commit/5d49073816f29129c0395f69648868f710754ced) Added nock recorder to all Jira http calls (@jwtd)

test/helper.js
'use strict';

// nodeunit
var util = require('util'),
    path = require('path'),
    grunt = require('grunt'),
    exec = require('child_process').exec;


// Duplicate the environment object
// NOTE: Environment variables in child processes are always strings
var envDup = [],
    envVar;
for (envVar in process.env) {
  if (process.env.hasOwnProperty(envVar)) {
    envDup[envVar] = process.env[envVar];
  }
}

// Prepare configuration for exec calls
var execOptions = {
  cwd: path.join(__dirname, '..'),  // Run in tests directory
  env: envDup                       // Pass in the duplicated env variables
  //encoding: 'utf8',
  //timeout: 0,            // kill child process if it runs longer than timeout milliseconds
  //maxBuffer: 200*1024,   // kill child process if data in stdout or stderr exceeds this limit
  //killSignal: 'SIGTERM'  // The child process is killed with killSignal (default: 'SIGTERM')
};


module.exports = {

  envDup: envDup,

  // Call grunt with the correct flags
  // Examples:   grunt addJiraComment:withoutContent_should_FAIL --env=TEST --no-color
  callGrunt: function (task, whenDoneCallback) {
    exec('grunt ' + task + ' --env=TEST --no-color', execOptions, whenDoneCallback);
  },

  parseTestOutput: function(s) {
    var n;
    var blocks = s.split('||||');
    for (n in blocks) {
      if (blocks.hasOwnProperty(n)) {
        console.log('** ' + n + ' **\n' + blocks[n]);
      }
    }
    //console.log('Inspect Object :: ' + util.inspect(blocks, {showHidden: false, depth: null}));
  },

  // Call grunt with the correct flags
  splitOutput: function (s) {
    return s.split('||||');
  }

};

test/require_helper.js
'use strict';
/**
 * Updates require path so that instrumented code can be loaded during test coverage analysis
 *
 * @param String path
 *    Any valid require path
 *
 * @return String path
 *    The path to the correct environment
 *
 * Using requireHelper in a test
 *    var requireHelper = require('../require-helper');
 *    var formValidator = requireHelper('form-validator');
 *
 */
module.exports = function (path) {
  return require((process.env.APP_DIR_FOR_CODE_COVERAGE || '../tasks/') + path);
};

Gruntfile.js
/*
 * grunt-jira-actions
 * https://github.com/jwtd/grunt-jira-actions
 *
 * Copyright (c) 2015 Jordan Duggan
 * Licensed under the MIT license.
 */

'use strict';

/*
 Jira API Docs
 https://docs.atlassian.com/jira/REST/latest/#d2e4023
*/

module.exports = function(grunt) {


  // Time grunt task execution
  require('time-grunt')(grunt);


  /*-----------------------------------------------------*
   *        Global properties for use in all tasks        *
   *------------------------------------------------------*/


  // Prepare to calculate paths
  var path = require('path');
  var root = path.resolve();

  // Create global config (gc)
  var gc = {
    env: grunt.option('env') || 'default',
    root: root,
    srcDir: 'src',
    docsSrcDir: 'docs',
    testsDir: 'test',
    configDir: 'config',
    buildDir: 'build'
  };


  // Project configuration.
  grunt.initConfig({


    // Pull in the package details
    pkg: grunt.file.readJSON('package.json'),

    // Common paths for our tasks to use
    gc: gc,


    /*--------------------------------*
     *        Code Conventions        *
     *--------------------------------*/


    eslint: {
      options: {
        config: 'eslint.json'
      },
      target: [
        'Gruntfile.js',
        'tasks/**/*.js',
        '<%= gc.testsDir %>/**/*_tests.js'
      ]
    },


    nodeunit: {
      all: ['<%= gc.testsDir %>/**/*_tests.js'],
      options: {
        reporter: 'verbose'
      },
      tap: {
        options: {
          reporter: 'tap',
          reporterOutput: '<%= gc.buildDir %>/reports/tap-test-results.tap',
          reporterOptions: {
            output: '<%= gc.buildDir %>/reports'
          }
        }
      },
      ci: {
        options: {
          reporter: 'junit', // Creates jUnit compatible XML reports, which can be used with continuous integration tools such as Hudson.
          reporterOutput: '<%= gc.buildDir %>/reports/junit-test-results.xml',
          reporterOptions: {
            output: '<%= gc.buildDir %>/reports'
          }
        }
      }
    },


    /*--------------------------------*
     *       Release Management        *
     *--------------------------------*/


    release: {
      options: {
        changelog: true, //default: false
        commitMessage: 'Release of <%= version %>',
        tagMessage: 'Version <%= version %>', //default: 'Version <%= version %>',
        //beforeBump: [],    // optional grunt tasks to run before file versions are bumped
        //afterBump: [],     // optional grunt tasks to run after file versions are bumped
        //beforeRelease: [], // optional grunt tasks to run after release version is bumped up but before release is packaged
        //afterRelease: [],  // optional grunt tasks to run after release is packaged
        github: {
          repo: '<%= pkg.repository.url.replace(\'https://github.com/\', \'\') %>', //put your user/repo here
          usernameVar: 'GITHUB_USERNAME', //ENVIRONMENT VARIABLE that contains Github username
          passwordVar: 'GITHUB_PASSWORD'  //ENVIRONMENT VARIABLE that contains Github password
        }
      }
    },

    // https://github.com/lalitkapoor/github-changes
    githubChanges: {
      dist: {
        options: {
          owner: 'jwtd',
          repository: 'grunt-jira-actions'
          //order-semver : use semantic Versioning for the ordering instead of the tag date
        }
      }
    },


    /*--------------------------------*
     *    Setup Jira Configuration    *
     *--------------------------------*/


    setJiraConfig: {
      options: {
        env_var_for_jira_username: 'JIRA_UN',
        env_var_for_jira_password: 'JIRA_PW',
        jira_host: 'virtru.atlassian.net',
        jira_protocol: 'https',
        jira_port: 443,
        jira_api_version: '2'
      }
    },


    /*--------------------------------*
     *         Add Jira Issues        *
     *--------------------------------*/


    // Create Jira issues
    createJiraIssue: {


      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'virtru.atlassian.net',
        project_id: 10400
      },


      /*- Pass Cases ------------------*/


      withMinimumOptions_should_PASS: {
        options: {
          summary: 'Issue from project_id, jira_host, summary, and defaults for everything else'}
      },


      asValidStoryDescrFromOption_should_PASS: {
        options: {
          issue_type: 'Story', // Story, Epic, Task, Technical Task, Sub-Task, Bug, Improvement, New Feature
          summary: 'Story with summary and description from a target option',
          description: 'This is the story description as string.'
        }
      },


      asValidTaskDescrFromFile_should_PASS: {
        options: {
          issue_type: 'Task', // Story, Epic, Task, Technical Task, Sub-Task, Bug, Improvement, New Feature
          summary: 'Task with summary and description from a filepath',
          description: 'test/data/issue_body.txt'
        }
      },


      asValidStoryMarkedDone_should_PASS: {
        options: {
          issue_state: 2,  // 1 = Open, 2 = Done
          summary: 'Story marked as done'
        }
      },


      asValidStoryWithPriority_should_PASS: {
        options: {
          summary: 'This is the bar task summary',
          additional_fields: {
            'priority': {
              'name': 'Major' // Critical, Major, Medium (default), Minor
            }
          }
        }
      },


      /*- Failure Cases ------------------*/


      withoutRequiredOptions_should_FAIL: {
        options: {
        }
      },


      withInvalidIssueType_should_FAIL: {
        options: {
          issue_type: 'BLARG', // Story, Epic, Task, Technical Task, Sub-Task, Bug, Improvement, New Feature
          summary: 'Should fail because BLARG is not a valid issue type'
        }
      },


      withInvalidIssueState_should_FAIL: {
        options: {
          issue_state: 42, // 1 = Open, 2 = Done
          summary: 'Should fail because 42 is not a valid issue state'
        }
      },


      withInvalidPriority_should_FAIL: {
        options: {
          summary: 'Should fail because BLARG is not a valid priority name',
          additional_fields: {
            'priority': {
              'name': 'BLARG' // Critical, Major, Medium (default), Minor
            }
          }
        }
      }

    },


    /*--------------------------------*
     *     Transition Jira Issue      *
     *--------------------------------*/


    transitionJiraIssue: {
      options: {
        jira_host: 'virtru.atlassian.net'
      }
    },


    /*--------------------------------*
     *       Link Jira Issues         *
     *--------------------------------*/


    // linkJiraIssue:GEN-200:Relates:GEN-201
    linkJiraIssue: {
      options: {
        jira_host: 'virtru.atlassian.net',
        from_issue_key: 'GEN-345',
        link_type: 'Relates',              // Blocks, Clones, Duplicate, Relates
        to_issue_key: 'GEN-123',
        comment: 'This is a test link'
      }
    },


    /*--------------------------------*
     *       Add Jira Comments        *
     *--------------------------------*/


    // Add comments to existing Jira issues
    addJiraComment: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'virtru.atlassian.net'
      },


      /*- Success Cases ------------------*/


      // Create specific targets to perform different Jira tasks
      fromOption_should_PASS: {
        options: {
          // issue_id: This value will be passed in via target call
          comment: 'This is the comment as a string.'
        }
      },

      // Create specific targets to perform different Jira tasks
      fromFileToIssue_should_PASS: {
        options: {
          // issue_id: This value will be passed in via target call such as addJiraComment:fromFileToIssue:19400
          comment: 'test/data/comment_body.txt'
        }
      },


      /*- Failure Cases ------------------*/


      // Create specific targets to perform different Jira tasks
      // grunt createJiraIssue:addJiraComment:withoutContent_should_FAIL --env=TEST --no-color
      withoutContent_should_FAIL: {
        options: {
        }
      },

      // Create specific targets to perform different Jira tasks
      withoutPassingIssueId_should_FAIL: {
        options: {
          // issue_id: This value will be passed in via target call
          comment: 'This should fail, because no issue id was passed in when it was called.'
        }
      }


    },


    /*--------------------------------*
     *    Create a Project Version    *
     *--------------------------------*/


    // Create a Version of a Jira project
    createJiraVersion: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'virtru.atlassian.net'
      },

      // Create specific targets for different Jira projects
      general: {
        options: {
          project_key: 'GEN',
          name: 'New Version 1',
          description: 'test/data/version_description.txt',
          archived: false,
          released: true,
          release_date: '2015-02-21'
          //userReleaseDate: '5/Jul/2010'
        }
      }
    },


    /*--------------------------------*
     *   Lookup Jira project details  *
     *--------------------------------*/


    // Lookup Jira project details
    jiraProjectDetails: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'virtru.atlassian.net'
      },

      // Create specific targets for different Jira project lookups
      forGeneralProject: {
        options: {
          project_key: 'GEN'
        }
      }
    },


    // Lookup Jira project rapid view
    jiraProjectRapidView: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'virtru.atlassian.net'
      },

      // Create specific targets for different Jira project lookups
      forGeneralProject: {
        options: {
          project_key: 'General'
        }
      }
    },


    /*--------------------------------*
     *     Search Jira using JQL      *
     *--------------------------------*/


    // Search Jira for specific issues
    searchJira: {

      // Declare options that are common to all Jira actions
      options: {
        jira_host: 'virtru.atlassian.net'
      },

      // Create specific targets for different Jira searches
      //    /rest/api/2/search?
      //      jql=
      //        project=WS
      //        +AND+status=%22OPEN%22
      //        +AND+issuetype+in%20(%22Bug%22,%22Story%22)
      forGenIssues: {
        options: {
          search_string: 'project="WS" AND status="OPEN" AND issuetype in ("Bug", "Story")'
          //start_at: 0,
          //max_results: 3,
          //fields: {},
          //before_search: [],   // optional grunt tasks to run before search
          //after_search: []     // optional grunt tasks to run after search
        }
      }
    }

  });


  // Load from package.json
  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  // Load this plugin's tasks
  grunt.loadTasks('tasks');


  // Project tasks
  grunt.registerTask('check', ['eslint', 'test']);
  grunt.registerTask('test', ['nodeunit']);
  grunt.registerTask('changelog', ['githubChanges']);

  grunt.registerTask('issue', ['setJiraConfig', 'createJiraIssue']);
  grunt.registerTask('transition', ['setJiraConfig', 'transitionJiraIssue']);
  grunt.registerTask('link', ['setJiraConfig', 'linkJiraIssue']);
  grunt.registerTask('comment', ['setJiraConfig', 'addJiraComment']);
  grunt.registerTask('version', ['setJiraConfig', 'createJiraVersion']);
  grunt.registerTask('project', ['jiraProjectDetails:forGeneralProject']);
  grunt.registerTask('project-view', ['jiraProjectRapidView:forGeneralProject']);
  grunt.registerTask('search', ['setJiraConfig', 'searchJira:forGenIssues']);

  grunt.registerTask('default', ['check']);

};
